File: problem4.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        // Make sure that left/first one is smaller one
        //  If nums1 has more number, we will switch nums1 and nums2
        if(nums1.size() > nums2.size()){
            vector<int> temp = nums1;
            nums1 = nums2;
            nums2 = temp;
        }

        // m and n are size of arrays
        int m = nums1.size();
        int n = nums2.size();

        // total is total number, and half is median index of total numner
        int total = m + n;
        int half = total / 2;

        // If total num is 0, return 0 
        // If total num is 1, return that number
        if(total == 0){
            return 0;
        }
        else if( total == 1){
            // Nums2 is almost bigger one
            return nums2[0];
        }
        // Starting index of each arrays if they are combined.
        float left = 0;
        float right = m;
        float nums1left = 0;
        float nums1right = 0;
        float nums2left = 0;
        float nums2right = 0;

        while(true){
            // Left is 0 and right is m initially, i -> m/2
            // j -> (m+n)/2 - (m)/2 -> n/2 initially
            int i = (left + right )/2; // Partition one
            int j = half - i; // Partition two

            if( i > 0){
                nums1left = nums1[i-1];
            }
            else{
                nums1left = -99999999;
            }

            if( i< m ){
                nums1right = nums1[i];
            }
            else{
                nums1right = 99999999;
            }

            if ( j > 0){
                nums2left = nums2[j-1];
            }
            else{
                nums2left = -999999999;
            }
            
            if(j < n){
                nums2right = nums2[j];
            }
            else{
                nums2right = 99999999;
            }

            if(nums1left <= nums2right && nums2left <= nums1right){
                if(total % 2 == 0){
                    float ans = (max(nums1left, nums2left) + min(nums1right, nums2right)) / 2;
                    cout << ans;
                    return ans;
                }
                else{
                    return min(nums1right, nums2right);
                }
            }
            else if (nums1left > nums2right){
                right = i - 1;
            }         
            else{
                left = i + 1;
            }
        }
    }
};

File: problem4.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]):
        # Index of two arrays
        index1 = 0
        index2 = 0
        totalNum = len(nums1) + len(nums2)
        newArray = []
        # Loop until all items into to newArray
        while len(newArray) < totalNum:
            # All numbers from nums1 in newArray, add remaining numbers from nums2
            if index1 == len(nums1) :
                newArray.append(nums2[index2])
                index2 += 1
            # All numbers from nums2 in newArray, add remaining numbers from nums1
            elif index2 == len(nums2):
                newArray.append(nums1[index1])  
                index1 += 1
            elif nums1[index1] < nums2[index2]:
                newArray.append(nums1[index1])
                index1 +=1
            else:
                newArray.append(nums2[index2])
                index2 += 1
                
        # Find the median and check if total number is even or odd.
        medIndex = len(newArray) // 2
        if len(newArray) % 2 == 0:
            return (newArray[medIndex-1] + newArray[medIndex])/2
        else:
            return newArray[medIndex]
